src/server/server.h

```diff
@@ -84,6 +88,24 @@ struct ChannelSubscribeNum {
   size_t subscribe_num;
 };
 
+// CURSOR_DICT_SIZE must be 2^n where n <= 16
+constexpr const size_t CURSOR_DICT_SIZE = 1024 * 16;
+static_assert((CURSOR_DICT_SIZE & (CURSOR_DICT_SIZE - 1)) == 0, "CURSOR_DICT_SIZE must be 2^n");
+static_assert(CURSOR_DICT_SIZE <= (1 << 15), "CURSOR_DICT_SIZE must be less than or equal to 2^16");
+
+enum CursorType {
+  kTypeBase,  // cursor for SCAN
+  kTypeHash,  // cursor for HSCAN
+  kTypeSet,   // cursor for SSCAN
+  kTypeZSet,  // cursor for ZSCAN
+};
+
+struct CursorDictElement {
+  uint64_t cursor;
+  CursorType cursor_type;
```

Comment:
Could this also be encoded to the cursor? Seems we can take about 4 bit out of 32bit hash for it.